source/extensions/filters/http/jwt_authn/verifier.cc

```diff
@@ -0,0 +1,163 @@
+#include "extensions/filters/http/jwt_authn/verifier.h"
+
+#include "common/common/enum_to_int.h"
+#include "common/http/message_impl.h"
+#include "common/http/utility.h"
+
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtProvider;
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtRequirement;
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtRequirementAndList;
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtRequirementOrList;
+using ::Envoy::Http::LowerCaseString;
+using ::google::jwt_verify::Status;
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace JwtAuthn {
+
+void AuthVerifier::beforeVerify(const std::vector<std::string>& audiences, Http::HeaderMap& headers,
+                                Verifier::Callbacks& callback) {
+  auth_ = factory_(audiences);
+  callback_ = &callback;
+  // Remove headers configured to pass payload
+  auth_->sanitizePayloadHeaders(headers);
+}
+
+void AuthVerifier::close() {
+  if (auth_) {
+    auth_->onDestroy();
+  }
+}
+
+ProviderNameVerifier::ProviderNameVerifier(
+    const std::string& provider_name, const std::vector<std::string>& audiences,
+    const AuthFactory& factory, const Protobuf::Map<ProtobufTypes::String, JwtProvider>& providers)
+    : AuthVerifier(factory), audiences_(audiences) {
+  const auto& it = providers.find(provider_name);
+  if (it == providers.end()) {
+    throw EnvoyException(fmt::format("Required provider ['{}'] is not configured.", provider_name));
+  }
+  for (const auto& header : it->second.from_headers()) {
+    extract_param_.header_keys_.insert(LowerCaseString(header.name()).get() +
+                                       header.value_prefix());
+  }
+  for (const auto& param : it->second.from_params()) {
+    extract_param_.param_keys_.insert(param);
+  }
+  if (it->second.from_headers_size() == 0 && it->second.from_params_size() == 0) {
+    extract_param_.header_keys_.insert("authorizationBearer ");
+  }
```

Comment:
I suggest to create an Extractor with specific headers_key and param_keys for each ProviderNameVerifier.

Reply:
I created an extractor for each verifier with issuer. Allow failed or missing is still uses the extractor in authenticator.